src/test/java/uk/gov/pay/directdebit/infra/PostgresContainer.java

```diff
@@ -20,72 +20,82 @@
+class PostgresContainer {
 
-    private static final Logger logger = LoggerFactory.getLogger(PostgresContainer.class);
-
-    private final String containerId;
-    private final int port;
-    private DockerClient docker;
-    private String host;
-    private volatile boolean stopped = false;
+    static final String DB_USERNAME = "postgres";
+    static final String DB_PASSWORD = "mysecretpassword";
 
+    private static final Logger LOG = LoggerFactory.getLogger(PostgresContainer.class);
     private static final String GOVUK_POSTGRES_IMAGE = "govukpay/postgres:9.4.4";
-    private static final String DB_PASSWORD = "mysecretpassword";
-    private static final String DB_USERNAME = "postgres";
-    private static final int DB_TIMEOUT_SEC = 15;
     private static final String INTERNAL_PORT = "5432";
+    private static final int DB_TIMEOUT_SEC = 15;
 
-    public PostgresContainer(DockerClient docker, String host) throws InterruptedException, IOException, ClassNotFoundException, DockerException {
-        Class.forName("org.postgresql.Driver");
+    private final String containerId;
+    private final DockerClient docker;
+    private final String postgresUri;
 
-        this.docker = docker;
-        this.host = host;
+    private volatile boolean stopped = false;
 
+    PostgresContainer(DockerClient docker, String host) throws Exception {
+        Class.forName("org.postgresql.Driver");
         failsafeDockerPull(docker, GOVUK_POSTGRES_IMAGE);
-        docker.listImages(DockerClient.ListImagesParam.create("name", GOVUK_POSTGRES_IMAGE));
-
-        final HostConfig hostConfig = HostConfig.builder().logConfig(LogConfig.create("json-file")).publishAllPorts(true).build();
-        ContainerConfig containerConfig = ContainerConfig.builder()
-                .image(GOVUK_POSTGRES_IMAGE)
-                .hostConfig(hostConfig)
-                .env("POSTGRES_USER=" + DB_USERNAME, "POSTGRES_PASSWORD=" + DB_PASSWORD)
-                .build();
-        containerId = docker.createContainer(containerConfig).id();
+        this.docker = docker;
+        this.containerId = createContainer(docker);
         docker.startContainer(containerId);
-        port = hostPortNumber(docker.inspectContainer(containerId));
+        this.postgresUri = "jdbc:postgresql://" + host + ":" + getContainerPortBy(docker, containerId) + "/";
         registerShutdownHook();
         waitForPostgresToStart();
     }
 
-    public String getUsername() {
-        return DB_USERNAME;
+    String getPostgresDbUri() {
+        return postgresUri;
+    }
+
+    void stop() {
+        if (!stopped) {
+            try {
+                stopped = true;
+                LOG.info("Killing postgres container with ID: " + containerId);
+                LogStream logs = docker.logs(containerId, DockerClient.LogsParam.stdout(), DockerClient.LogsParam.stderr());
+                LOG.info("Killed container logs:\n");
+                logs.attach(System.out, System.err);
```

Comment:
I would've expected us to do a `LOG.info(logs.readFully())` or something, if we wants the logs to show up. What are we doing with the docker logs, instead?

Reply:
I'll have a look and raise a new PR for this ... need to investigate a bit

Reply:
(Y)